### PR TITLE
add spec/windows.dd

### PR DESCRIPTION
**Pull request by WalterBright**
The idea is to provide much better information about how Windows executables and dlls work and are constructed.

This is just a skeleton outline of the future page.

### CODE REVIEW

File: spec/windows.dd [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+$(P This covers Windows programming with 32 bit OMF, 32 bit Mscoff,
+64 bit Mscoff, console programs, GUI programs, and DLLs.
+)
+
+$(H2 $(LNAME2 mscoff, Windows 32 and 64 bit MSCOFF Programs))
+
+    $(P 32 bit and 64 bit MSCOFF programs use the Microsoft Visual C/C++ compiler as the Associated
+    C Compiler, generate object files in the MSCOFF format, and use the Microsoft linker
+    to link them.
+    )
```
Comment (rikkimax) [inline spec/windows.dd:16]: That'll need to cover mingw and lld (which you can't enable if you have VS).

File: spec/windows.dd [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    as the Associated C Compiler, generate object files in the OMF format, and use the
+    $(LINK2 https://www.digitalmars.com/ctg/optlink.html, Optlink linker)
+    to link them.
+    )
+
+$(H3 $(LNAME2 omf-console, Console Programs))
+
+$(H3 $(LNAME2 omf-windows, Windows GUI Programs))
+
+$(H3 $(LNAME2 omf-dlls, DLLs))
```
Comment (rikkimax) [inline spec/windows.dd:37]: That'll need to cover linker scripts which (for some reason) are required to generat a dll (at least last time I tried -shared wasn't enough).

**Comment by rikkimax** [comment]
There should also be some talk on DllMain and how the compiler won't automatically add it for you and the implementation required to for use with druntime.

File: spec/windows.dd (hunk 1 above)
Comment (WalterBright) [reply spec/windows.dd:16]: I don't know the ecosystem around mingw/lld. Someone who does will have to write it.

File: spec/windows.dd (hunk 2 above)
Comment (WalterBright) [reply spec/windows.dd:37]: I know. This page is currently just an outline.

File: spec/windows.dd (hunk 1 above)
Comment (rikkimax) [reply spec/windows.dd:16]: > 32 bit and 64 bit MSCOFF programs use the Microsoft Visual C/C++ compiler as the Associated C Compiler, generate object files in the MSCOFF format, and use the Microsoft linker if available to link them. If not available mingw import libraries and LLD linker will be used instead.

Shouldn't need much, like what I wrote above.

File: spec/windows.dd (hunk 2 above)
Comment (rikkimax) [reply spec/windows.dd:37]: All good, just noting it from experience :)

File: spec/windows.dd (hunk 1 above)
Comment (mdparker) [reply spec/windows.dd:16]: That should specify that it's the MinGW import libraries and LLD that ship with dmd. Anyone attempting to use the import libraries in the MinGW distro may encounter difficulties, so it's important to clarify that.

**Comment by ibuclaw** [comment]
What part of this is to do with the D language specification?

**Comment by adamdruppe** [comment]
yeah i agree, this has nothing to do with the D language. might be a nice addition under dmd's documentation though.

**Comment by WalterBright** [comment]
> What part of this is to do with the D language specification?

The same reason we have an ABI section.

**Comment by WalterBright** [comment]
And a C portability guide, and a C++ portability guide, etc. If you want do split all this non-spec stuff off into a separate subdirectory, I'm all for it. Please do it! Meanwhile, my focus is on solving the problem of mass confusion about how things work under the hood.

**Comment by WalterBright** [comment]
ping @dkorpel @RazvanN7 @maxhaton

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**